### PR TITLE
Implement auto-fix functionality

### PR DESCRIPTION
**Pull request by dar5hak**
I have added a mechanism to optionally instruct vala-lint to fix mistakes which are easily fixable.

## Implementation

Every check which is eligible for auto-fix overrides a method called `apply_fix`. This method receives the contents of the file and mutates them to apply the respective fix.

I have added a class called `Fixer`, which comes into play when the user passes the `--fix` or `-f` flag. It iterates through all the non-disabled mistakes and calls `apply_fix` for each of them.

With the updated contents, it writes the file and prints any remaining mistakes which could not be auto-fixed.

## Fixes

At the moment, the following mistakes get fixed:

- `trailing-whitespace`
- `trailing-newlines`
- `space-before-paren`
- `double-semicolon`

## Pending work

- Other checks
- Tests
- ~Multiple, clashing mistakes: The four aforementioned fixes happen to play well with one another, but I can imagine how fixing one mistake could alter the begin and end locations of other mistakes. I haven't found a way to deal with that.~

Can you review the current code and see if the mechanism makes sense? Are there any other potential problems I might have overlooked?

### CODE REVIEW

**Comment by dar5hak** [comment]
In fact, I just noticed that if there are more than one `space-before-paren` mistakes on the same line, only the first one gets fixed correctly. For subsequent ones, it adds a space at the wrong position, because the begin and end have changed.

One way of solving this would be modifying `apply_fix` to return exactly where it changed the contents and whether it was an addition or a deletion. Then, we could use that information to update the begin and end position of the remaining mistakes. Thoughts?

**Comment by jonafato** [comment]
A few ideas on how to address this class of issue:

1. Apply fixes in reverse (end of line to front of line). This should prevent earlier fixes from affecting later ones because of shifting contents. Most likely the easiest solution to implement.
2. Copying the contents to reference error positions, then rebuilding the output from the original or modified source, depending on the error status of any given position.
3. The "safest" / most stable approach is probably to read the source into a syntax tree and re-write it with consistent styling rules applied. I think that's beyond the scope of this proposal, but it seems like the option least likely to encounter edge cases like this, as well as fixes that interfere with each other / introduce new errors (e.g. adding a missing space before a paren, thereby pushing a line over the length limit).

Two additional questions:

1. Python's [`black`](https://github.com/psf/black#note-this-is-a-beta-product) formatter compares the AST of the original code to the AST of the modified code to ensure that the actual behavior hasn't changed as a result of reformatting. Is something like this possible here?
2. Pending work mentions clashing mistakes. Would this include an option to run the fixer multiple in multiple iterations to address the case of an applied fix introducing another fixable error?
3. If a new error is introduced by an applied fix (like adding a space in my example above), will that be reported in the output after applying fixes, or will it require another run to detect errors that require manual work?

(Not a contributor to the project, just a user of it that would love to see this feature, so thanks for putting the work in here!. Sharing my thoughts in case they're helpful.)

**Comment by dar5hak** [comment]
@jonafato Thanks a lot for your helpful inputs.

Yes, applying the fixes in reverse is the simplest solution, and should do the job. The only issue I can imagine is if there are mistakes with overlapping ranges. Haven't seen anything of that sort yet, will have to check.

> Copying the contents to reference error positions, then rebuilding the output from the original or modified source, depending on the error status of any given position.

I don't quite understand this approach. Could you elaborate?

And as you said, checking an AST is indeed out of the scope of this change. We mainly just point and shoot at individual mistakes that the lint step has already found.

That said, building an opinionated formatter (without the linting part) similar to Black and [Prettier](https://prettier.io/) would make a great project, but would likely be separate from vala-lint.

> Pending work mentions clashing mistakes.

What I meant here was when a fix would change the positions of other existing mistakes, solutions for which you already suggested. I cannot think of a scenario where a fix would introduce another mistake altogether. So another pass at finding more mistakes should not be needed.

**Comment by jonafato** [comment]
> > Copying the contents to reference error positions, then rebuilding the output from the original or modified source, depending on the error status of any given position.

> I don't quite understand this approach. Could you elaborate?

I wasn't very clear here. What I meant was creating a new string that data gets copied over from the original source into. For any positions where fixes wouldn't be applied, it just gets copied verbatim. If fixes would be applied, the fix functions could return the new string. I could go into more detail here, but it's most likely more effort than it's worth.

> I cannot think of a scenario where a fix would introduce another mistake altogether.

The only scenario I can think of here is one or more added spaces (from `space-before-paren`) pushing a line that's close to the length limit over it. I think this would result in a case where a new linting error is introduced for line length but not reported in the initial output.

---

One additional question: could this solution include a `--dry-run` option to display what changes would be made before actually making them? This would allow for some manual review before applying automatic fixes in the case of some sort of interaction between errors (e.g. if the fixers were to add a space in the middle of a variable because two rules overlapped).

**Comment by dar5hak** [comment]
> The only scenario I can think of here is one or more added spaces (from space-before-paren) pushing a line that's close to the length limit over it.

Good point. Breaking down a long line at a logical position isn't something I'm planning for this PR, but the least we could do here is add that to the mistakes list and print it for the user. I don't think that's great UX though, so ideas are welcome.

> could this solution include a --dry-run option to display what changes would be made before actually making them?

Yes. I thought of this too in #140. Let me see if it fits well in this PR or in a follow-up one, but it's definitely something I want to do.

**Comment by ChildishGiant** [comment]
For me the fix flag doesn't seem to do anything. There's no output and no files are changed.

**Comment by dar5hak** [comment]
@ChildishGiant Did you pass the relative paths to files? Because I fell into that trap too.

**Comment by ChildishGiant** [comment]
Ah yeah, now it works. I would have thought that it'd scan and run on all files that have fixable issues. I've also found that the `space-before-paren` rule is adding a space in the wrong place. 
`example(null)` -> `example( null)` instead of `example (null)`

**Comment by dar5hak** [comment]
Can you share the file? I'll debug it.

**Comment by ChildishGiant** [comment]
[example.vala.txt](https://github.com/vala-lang/vala-lint/files/7258169/example.vala.txt) Literally just that line reproduces the issue haha

**Comment by dar5hak** [comment]
This is not a bug in the fixing logic. The linter is detecting the error at the wrong position: it should be at position 1.7 (just before the parenthesis) but it reports it at 1.8. Oddly, the same works correctly for https://github.com/vala-lang/vala-lint/blob/master/test/files/space-before-paren-check.vala.

**Comment by vjr** [comment]
Could it be the presence of unicode BOM byte order mark at the beginning of the file causing the position to be off by one?

**Comment by dar5hak** [comment]
@vjr I was thinking on similar lines and looking for a zero-width space, but found none. But a BOM could be it. Do you know how to detect one?

**Comment by vjr** [comment]
I don't know offhand how to detect it... maybe compare the `GLib.FileInfo.get_size ()` with the linter position at the end of the `FileData`?

**Comment by vjr** [comment]
Maybe compare the length of `string content ` here:

https://github.com/vala-lang/vala-lint/blob/8baf751d2393f7766d66c77ba8e785d7af7c47c6/lib/Linter.vala#L79

Against the file size?

**Comment by dar5hak** [comment]
@vjr Yes, that sounds good, thanks! Might even need to check whether the first byte is indeed the BOM.

File: src/Application.vala [hunk 1]
```diff
@@ -44,6 +45,8 @@ public class ValaLint.Application : GLib.Application {
             "Always return a 0 (non-error) status code, even if lint errors are found." },
         { "generate-config", 'g', 0, OptionArg.NONE, ref generate_config_file,
             "Generate a sample configuration file with default values." },
+        { "fix", 'f', 0, OptionArg.NONE, ref auto_fix,
+            "Fix any auto-fixable mistakes." },
```
Comment (kgrubb) [inline src/Application.vala:49]: Can we add an entry in the README.md as well on how to use the flag, similar to this cli help text?

Comment (dar5hak) [reply src/Application.vala:49]: Good point. Added.

**Comment by dar5hak** [comment]
Thanks a lot, Keli. I'll add support for more rules as and when possible.

**Pull request merged**